packages/react-instantsearch/src/connectors/connectMenu.js

```diff
@@ -77,19 +77,37 @@ export default createConnector({
       return null;
     }
 
-    const items = results
-      .getFacetValues(attributeName, {sortBy})
-      .slice(0, limit)
-      .map(v => ({
-        value: getValue(v.name, props, searchState),
-        label: v.name,
-        count: v.count,
-        isRefined: v.isRefined,
-      }));
+    const items = searchForFacetValuesResults && searchForFacetValuesResults[attributeName]
+      ? searchForFacetValuesResults[attributeName]
+        .map(
+          v => ({
+            label: v.value,
+            value: getValue(v.value, props, searchState),
+            _highlightResult: {label: {value: v.highlighted}},
+            count: v.count,
+            isRefined: v.isRefined,
+          }))
+      : results
+        .getFacetValues(attributeName, {sortBy})
+        .map(v => ({
+          label: v.name,
+          value: getValue(v.name, props, searchState),
+          count: v.count,
+          isRefined: v.isRefined,
+        }));
+
+    const isFromSearch = Boolean(searchForFacetValuesResults && searchForFacetValuesResults[attributeName]);
+    const searchForFacetValues = props.searchForFacetValues ? this.searchForFacetValues : undefined;
+
+    const sortedItems = !isFromSearch && props.searchForFacetValues ?
+      orderBy(items, ['isRefined', 'count', 'label'], ['desc', 'desc', 'asc']) : items;
+    const transformedItems = props.transformItems ? props.transformItems(sortedItems) : sortedItems;
 
     return {
-      items,
+      items: transformedItems.slice(0, limit),
```

Comment:
Since inside transform items we allow users to completely change the items, I would say once the user provides a transformItems method we do not slice based on the limit and we warn that using limit AND transform items is not compatible.

Otherwise we will have usages like:

```
RefinementList
  transformItems=...
  limit=Infinity
```

Plus the fact that it does not make sense to allow the user doing a full filter and map on the whole items to then ditch items he mapped (If this is not clear, yell at me) :D